cvmfs/cache_ram.h

```diff
@@ -0,0 +1,191 @@
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual void CtrlTxn(const std::string &description,
+                       const ObjectType type,
+                       const int flags,
+                       void *txn);
+
+  /**
+   * Copy the given memory region into the transaction buffer. Copying starts at the transaction's current offset
+   * @param buf The source address
+   * @param size The number of bytes to copy
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual int64_t Write(const void *buf, uint64_t size, void *txn);
+
+  /**
+   * Seek to the beginning of the transaction buffer
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual int Reset(void *txn);
+
+  /**
+   * Commit a transaction and open the resulting cache entry. This is necessary to avoid a race condition in which the
+   * cache entry is evicted between the calls to CommitTxn and Open.
+   * @param txn A pointer to space allocated for storing the transaction details
+   * @returns A file descriptor to the new cache entry, or a negative error from @ref CommitTxn
+   */
+  virtual int OpenFromTxn(void *txn);
+
+  /**
+   * Free the resources allocated for the pending transaction
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual int AbortTxn(void *txn);
+
+   /**
+   * Commit a transaction to the cache. If there is not enough free space in the cache, first try to make room by evicting
+   * volatile entries. If there is still not enough room, try evicting regular entries. If there is *still* not enough
+   * space, give up an return failure. Note that evictions only occur if they will produce enough space for the transaction;
+   * if -ENOSPC is returned, the states of the transaction and cache are unchanged.
+   * @param txn A pointer to space allocated for storing the transaction details
+   * @returns -ENOSPC if the transaction would exceed the size of the cache
+   */
+  virtual int CommitTxn(void *txn);
+
+ private:
+  struct ReadOnlyFd {
+    ReadOnlyFd() : handle(shash::Any()), pos(0) { }
+    ReadOnlyFd(const shash::Any &h, uint64_t pos) : handle(h), pos(pos) { }
+    shash::Any handle;
+    uint64_t pos;
+  };
+
+  struct Transaction {
+    Transaction() :
+      buffer(NULL), size(0), expected_size(0), pos(0), object_type(kTypeRegular) { }
+    shash::Any id;
+    void *buffer;
+    uint64_t size;
+    uint64_t expected_size;
+    uint64_t pos;
+    ObjectType object_type;
+    std::string description;
+  };
+
+  inline bool IsValid(int fd) {
+    if ((fd < 0) || (static_cast<unsigned>(fd) >= open_fds_.size()))
+      return false;
+    return open_fds_[fd].handle != invalid_fd_;
+  }
+
+  int AddFd(const ReadOnlyFd &fd);
+  int64_t CommitToKvStore(Transaction *transaction);
+  virtual int DoOpen(const shash::Any &id);
+
+  uint64_t max_size;
+  shash::Any invalid_fd_;
```

Comment:
That can probably become a `static const`.  I think it would be better to use the term handle instead of fd, e.g. `kInvalidHandle`.